RFCs/0031-Multi-Test-Run-Finalization.md

```diff
@@ -0,0 +1,304 @@
+
+        /// <summary>
+        /// Dispatch MultiTestRunFinalizationProgress event to listeners.
+        /// </summary>
+        /// <param name="finalizationProgressEventArgs">Finalization Progress event args.</param>
+        void HandleMultiTestRunFinalizationProgress(MultiTestRunFinalizationProgressEventArgs finalizationProgressEventArgs);
+    }
+}
+```
+Interface provides callbacks from Multi Test Run Finalization process. For every such process `HandleMultiTestRunFinalizationComplete` will be called once and will provide last chunk or all data collector attachments. During finalization process `HandleFinalisedAttachments` can be invoked several times providing data collector attachments that are already processed. Method `HandleMultiTestRunFinalizationProgress` will be invoked every time when `progressReporter` is used by `IDataCollectorAttachmentProcessor` implementation and will provide information about current collector: progress, uris and index of collector. Additionally event will contain also number of processors.
+
+
+4. Use above logic to reprocess data collector attachments for parallel test executions and VS scenarios (e.g. `Run All Tests`, `Analyze Code Coverage for All Tests`). In case of `Analyze Code Coverage for All Tests` VS will use `vstest.console` in a variation of design mode and merge all code coverage reports. VS will show full code coverage report for all test projects.
+
+5. When [dotnet test](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-test) command is used to execute unit tests in a given solution, new console app `Orchestrator` will be executed.
+  * `Orchestrator` for every project inside solution will start `dotnet test` command, using `Process` with output redirected. Output for every project will be printed to stdout with some mutex to not mix output from children. 
+  * Additionally new test platform extension will be introduced and will propagate data about:
+    - Attachments
+    - Test results statistics
+  back to `Orchestrator`.
+  * In parallel with test executions `Orchestrator` will start `vstest.console` in Design Mode.
+  * Whenever at least 2 test executions are finished `Orchestrator` will invoke `FinalizeMultiTestRunAsync` and provide all attachments from those test executions that finished. Parameter `multiTestRunCompleted` will be set to `false`. Test platform will provide data collector attachments only to associative processors.
+  * When all test exections are done `Orchestrator` will provide all attachments back through `FinalizeMultiTestRunAsync` with `multiTestRunCompleted` set to `true`. Test Platform will use all available processors to process data collector attachments.
+  * When all attachments are merged `Orchestrator` will display information about data collector attachments to standard output. 
+  * Finally `Orchestator` will combine all tests results statistics and print it to standard output.
+
+![dotnet-test finalization](Images/finalization.png)
```

Comment:
>dotnet-test finalization](Images/finalization.png [](start = 2, length = 49)

Thanks, this makes things a lot more clearer. Few comments on the image:
1. Aren't the ports for each dotnet test process different? 
2. It looks like the communication between vstest.console and the orchestrator need to be updated. Should the arrows point in the reverse direction? The finalization also happens in the orchestrator doesn't it?
3. Can we indicate that intermediate test results are still sent back to the parent host?
4. When we say vstest.console, I'm assuming we are referring to the parent vstest/test task that would spawn off dotnet vstest/test with the orchestrator. We could specify it as so.

Reply:
The main process is second thing (Orchestrator). It made it not clear. I will fix it

Reply:
Do we need intermediate test results? I was thinking of taking all output from inner process and print it to stdout.

Reply:
I meant some indication that stdout would still be populated with test results in real time. We do not need individual test cases/results specially for this.

---
In reply to: [447198580](https://github.com/microsoft/vstest-docs/pull/235#discussion_r447198580) [](ancestors = 447198580)

Reply:
Added this in arrows when starting processes